src/util.jl

```diff
@@ -14,20 +14,21 @@ end
 
 # Base gives up on tuples for promote_eltype... (TODO can we improve Base?)
 @generated function promote_tuple_eltype{T <: Tuple}(::Union{T,Type{T}})
-    t = Union{}
+    types = []
     for i = 1:length(T.parameters)
         tmp = T.parameters[i]
         if tmp <: Vararg
             tmp = tmp.parameters[1]
         end
-        t = promote_type(t, tmp)
+        push!(types, tmp)
     end
     return quote
         $(Expr(:meta,:pure))
```

Comment:
It's what the `@pure` macro does. `@pure function f(x); ...; end` transforms to `function f(x); $(Expr(:meta, :pure)); ...; end` (similarly for `@inline`, `@noinline` and `@propagate_inbounds`). A better style is to use a `@_pure_meta` macro call here (which just inserts `$(Expr(:meta, :pure))`).

However, I've replaced this with `@_inline_meta` here since pure is the wrong concept for this.